### PR TITLE
Show requested version in Solution PMUI

### LINKED ISSUE #9827: Show requested version in the Solution Level UI
Currently we only show the installed version in the Solution Level, we are going to add a column "Requested" version. 
It only applies to PackageReference projects.

### PR DESCRIPTION
**Pull request by martinrrm**
## Bug

Fixes: [Show requested version in the Solution Level UI](https://github.com/nuget/home/issues/9827)
Regression: No

## Fix

Details:

@donnie-msft comments:

- I've refactored this a bit so that the GridViewColumn is dynamically added, rather than using a 0-width. I feel this is better for code readability, perf, and it does work better with Narrator (ie, it indicates correct # columns). 
  - Creating column when DataContext is changed.
- Simplified logic since `IsRequestedVisible` only needs to be set once, at PMUI creation time.
- Sorting was a little off due to the way user settings get restored. Explained in comments that I'm always creating a column header for Requested "Version", that way if the previous sort setting was this column, it can be restored during control initialization.
  - (well, actually, I found this doesn't work quite right for any column other than Project Name (https://github.com/NuGet/Home/issues/10054) .... but at least I've hooked it up correctly on this new column)
-  Renamed string to ColumnHeader_Requested to be more specific

@martinrrm comments:
Add a "Requested" column in the Solution PMUI

image
image
image
image

## Testing/Validation

Tests Added: No  
Reason for not adding tests:  [@donnie-msft]'s thought is this is mostly a plumbing of a property to the UI. The UI is simply creating a column when necessary. The effort doesn't seem valuable given the minimal impact of the change. Open to counterpoints.

Validation:  Open Solution PMUI, with a PC and PR project. Ensure the column shows/hides as expected.
Open Project PMUI... make sure not broken.

### CODE REVIEW

File: src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Shipped.txt
Comment (zivkan) [inline src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Shipped.txt]: Granted this is in the only assembly we're more liberal in allowing breaking changes. But in general, you should not be changing shipped APIs. Instead, create a new overload, and change the existing method to call the new overload with an appropriate default value for the new parameter.

Any customers using our package in their own app and using this API will be broken if this change ships as-is.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/BooleanToWidthConverter.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/BooleanToWidthConverter.cs]: license

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: Could this boolean be set inside your loop when setting RequestedVersion? 
Or is it possible to have a `null` installedVersion but that project still should show Requested versions?

File: src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Unshipped.txt
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Unshipped.txt]: this is a NoOp?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs]: Rename to Unloaded

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs]: Awesome! So after this event, when the PMUI reappears, the width looks good? And if you expand/reduce other column widths, nothing weird happens?

File: src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Unshipped.txt
Comment (martinrrm) [reply src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Unshipped.txt]: This is not mine, I think I saved this file while rebasing and it added a newline at the end

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (martinrrm) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: It should show the Requested versions column, this column depends on the project type, but I can move this inside the loop

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs
Comment (martinrrm) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs]: Yeah it works, but I erased the wrong function and pushed haha, it should be only `_versionColumn.Width` because in the header we are only changing the Visibility property

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: I'd use `_isRequestedVisible ` here so that we don't fire PropertyChanged unnecessarily.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: In my earlier suggestion, I was thinking to simply set IsRequestedVisible inside of the `if` statement you already have here: https://github.com/NuGet/NuGet.Client/pull/3555/commits/bfd4f346c9c377deeab4ac8b2098aa6aea538855#diff-17f11a93b3135e66bd31d0d723de815aR99

Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: I'd use either `!IsRequested` or `== false` since `.Equals(false)` feels a bit weird.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: Better yet... You could use a local like `bool anyRequested = false`, then set it once down here https://github.com/NuGet/NuGet.Client/pull/3555/commits/bfd4f346c9c377deeab4ac8b2098aa6aea538855#diff-17f11a93b3135e66bd31d0d723de815aR122 (eg, `IsRequestedVisible = anyRequested`)

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: use local here if you agreed with my suggestion

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: I suggested another refactoring. I meant that your loop already does this work, so not sure we need to call `.Any` again?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: Was the decision to display the request column whenever a PR project is available? 
I thought it was only when *all* projects are PR.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: use `using NuGet.ProjectModel`.
Don't fully qualify the enum

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: I added a PR to improve our guidelines with this: https://github.com/NuGet/NuGet.Client/pull/3592

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: You are already doing a project style check above right?

File: src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Unshipped.txt
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Unshipped.txt]: If that's the case, I think you should revert this. 

Sample command: 

> git checkout dev -- src/NuGet.Core/NuGet.PackageManagement/PublicAPI.Unshipped.txt 

Should revert only that specific file.

**Review by nkolev92 — COMMENTED** [review]
Can you please add screenshots for the relevant scenarios? 

personally I'd like to see example for: 

* PR project level.
* PR only solution
* PC only solution
* PR + PC mixed solution

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: I think you're asking what I asked here https://github.com/NuGet/NuGet.Client/pull/3555#discussion_r473109979
and now here https://github.com/NuGet/NuGet.Client/pull/3555#discussion_r473268210

Comment (martinrrm) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: Yes, but this is for the package, here we are comparing if the package installed is from a PR project, otherwise the package doesn't have a requested version, because PC project don't use Allowed Version

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (martinrrm) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: The link is not taking to a specific line of code, but I think you are referring to the if below where I check the project type, that is inside a try so it can fail

Comment (martinrrm) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs]: I'm going to add an image to be more explicit, but you can have the requested column even though you don't have package with a requested version or installed, this column is going to show if there is a PR project so we don't confuse users on when this column is being show. @nkolev92

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/BooleanToWidthConverter.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/BooleanToWidthConverter.cs]: Have you had a chance to test the narrator here? 
https://github.com/NuGet/NuGet.Client/blob/dev/docs/feature-guide.md#visual-studio-ui-considerations

**Review by nkolev92 — DISMISSED** [review]
I think you should test the narrator to make sure the 0-width column does not get announced.
Looks great beyond that. 

Only thing I'm concerned is that the 0 width column might get announced.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/BooleanToWidthConverter.cs
Comment (martinrrm) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/BooleanToWidthConverter.cs]: Yeah it is getting announced when the width is 0, the narrator says 4 columns

Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/BooleanToWidthConverter.cs]: Hmm... would we want to allow screen readers to indicate the column is blank? Or should the column be logically removed?

Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/BooleanToWidthConverter.cs]: I completely refactored this and Narrator indicates correct # columns now.

**Comment by donnie-msft** [comment]
@rrelyea please take a look at description for my comments on this PR.
@nkolev92 dismissed your review as I've made a lot of changes here.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:137]: How about `if (IsRequestedVisible)` ? 

```suggestion
                        if (IsRequestedVisible)
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs]: No need to change it, but `is` has an optimization so you can write something like:

```cs
if ( e.NewValue is PackageSolutionDetailControlModel model)
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs:260]: I find this weird, but sounds like it solves the narrator problem. 
I'd wait on a review from someone skilled with xaml stuff.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs]: how did we come up with a diff of 7? :D

**Review by nkolev92 — APPROVED** [review]
Looks fine overall to me. 

Couple of small questions. 
I'd prefer if someone else approved the requested column addition trick. 
I am happy that it solves the problem, and don't have any particular ideas, but it increases the code complexity in a way that's not obvious.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs]: It's arbitrary, but the word Installed is a tiny bit wider than Version, so that's why I expanded

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs:260]: @rrelyea & @srdjanjovcic

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs]: You inspired me to make this vary based on column name. I also shrunk to 64 since it still looks good at that spacing.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:137]: Can't some projects be PR and others be PC? I thought this was project-by-project determination

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs]: Did it since was making other changes

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:137]: You are right, it was a different method from the above one.
I'm pretty sure I provided the same feedback to Martin :) 

Resolving this.

**Pull request merged**
